File: Trees/131-palindrome-partitioning/palindrome-partitioning.cpp
class Solution {
public:
    vector<vector<string>> partition(string s) {
        
        vector<vector<string>> ans;

        vector<string> temp;

        findpartition(s, ans, 0, temp);
        
        return ans;
    }

    void findpartition(string s, vector<vector<string>>& ans, int start, vector<string>& temp)
    {
        if (start == s.size())
        {
            ans.push_back(temp);
            return;
        }

        for (int i = start; i < s.size(); i++)
        {
            if (isPalindrome(s, start, i))
            {
                temp.push_back(s.substr(start, i - start + 1));
                findpartition(s, ans, i + 1, temp);
                temp.pop_back();
            }
        }
    }

    bool isPalindrome(string s, int start, int end) {
        while (start <= end)
        {
            if (s[start++] != s[end--])
            {
                return false;
            }
        }

        return true;
    }
};

File: DP/131-palindrome-partitioning/palindrome-partitioning.java
class Solution {

    private static void recur(String s,int currIndex,List<String> curr,List<List<String>> ans) {

        if(currIndex == s.length()) {
            ans.add(new ArrayList<>(curr));
            return ;
        }

        for(int i = currIndex ; i < s.length() ; i++) {
            if(isPalindrome(s,currIndex,i)) {

                curr.add(s.substring(currIndex,i + 1));

                recur(s,i + 1,curr,ans);

                curr.remove(curr.size() - 1);
            }
        }

    }

    private static boolean isPalindrome(String s,int i,int j) {
        while(i <= j) {
            if(s.charAt(i) != s.charAt(j)) return false;
            i++;
            j--;
        }

        return true;

    }

    public List<List<String>> partition(String s) {
        List<List<String>> ans = new ArrayList<>();

        List<String> curr = new ArrayList<>();

        recur(s,0,curr,ans);

        return ans;
    }
}